File: src/v3/cache.test.ts
import { describe, expect, test } from 'vitest';
import { WeightedLruCache } from './cache.js';

describe('weighted LRU cache', () => {
  test('promotes hits and evicts the least-recently-used entry by count', () => {
    const cache = new WeightedLruCache<string, number>(2, 10);
    cache.set('first', 1, 1);
    cache.set('second', 2, 1);
    expect(cache.get('first')).toBe(1);
    cache.set('third', 3, 1);

    expect(cache.get('second')).toBeUndefined();
    expect(cache.get('first')).toBe(1);
    expect(cache.get('third')).toBe(3);
    cache.set('third', 30, 2);
    expect(cache.get('third')).toBe(30);
    expect(cache.stats()).toMatchObject({ entries: 2, weight: 3, evictions: 1 });
  });

  test('enforces aggregate weight and never retains an oversized entry', () => {
    const cache = new WeightedLruCache<string, number>(10, 4);
    cache.set('first', 1, 3);
    cache.set('second', 2, 2);
    expect(cache.get('first')).toBeUndefined();
    expect(cache.stats()).toMatchObject({ entries: 1, weight: 2, evictions: 1 });

    cache.set('oversized', 3, 5);
    expect(cache.get('oversized')).toBeUndefined();
    expect(cache.stats()).toMatchObject({ entries: 1, weight: 2 });
  });

  test('clear removes entries and resets counters', () => {
    const cache = new WeightedLruCache<string, number>(1, 1);
    cache.get('missing');
    cache.set('value', 1);
    cache.get('value');
    cache.clear();
    expect(cache.stats()).toEqual({
      entries: 0,
      weight: 0,
      hits: 0,
      misses: 0,
      evictions: 0,
    });
  });

  test('stops eviction safely if constructed with an impossible internal cap', () => {
    const cache = new WeightedLruCache<string, number>(-1, 1);
    cache.set('value', 1);
    expect(cache.stats()).toMatchObject({ entries: 0, weight: 0 });
  });
});

File: go/cache_test.go
package dicecore

import "testing"

func TestWeightedCacheEvictionAndPromotion(t *testing.T) {
	c := NewWeightedLruCache[string, int](2, 10)
	c.Set("first", 1)
	c.Set("second", 2)
	if v, ok := c.Get("first"); !ok || v != 1 {
		t.Fatal(v, ok)
	}
	c.Set("third", 3)
	if _, ok := c.Get("second"); ok {
		t.Fatal("least recent entry retained")
	}
	c.Set("third", 30, 2)
	if v, ok := c.Get("third"); !ok || v != 30 {
		t.Fatal(v, ok)
	}
	if got := c.Stats(); got.Entries != 2 || got.Weight != 3 || got.Hits != 2 || got.Misses != 1 || got.Evictions != 1 {
		t.Fatal(got)
	}
	c.Clear()
	if got := c.Stats(); got != (LruCacheStats{}) {
		t.Fatal(got)
	}
}

func TestWeightedCacheCaps(t *testing.T) {
	c := NewWeightedLruCache[string, int](10, 4)
	c.Set("first", 1, 3)
	c.Set("second", 2, 2)
	if _, ok := c.Get("first"); ok {
		t.Fatal("weight cap ignored")
	}
	c.Set("oversized", 3, 5)
	if _, ok := c.Get("oversized"); ok {
		t.Fatal("oversized entry retained")
	}
	c.Set("second", 4, 5)
	if got := c.Stats(); got.Entries != 0 || got.Weight != 0 || got.Evictions != 1 {
		t.Fatal(got)
	}
	for _, caps := range [][2]int64{{0, 1}, {1, 0}, {-1, 1}, {1, -1}} {
		cache := NewWeightedLruCache[string, int](caps[0], caps[1])
		cache.Set("x", 1)
		if got := cache.Stats(); got.Entries != 0 || got.Weight != 0 {
			t.Fatal(got)
		}
	}
}
